internal/component/database_observability/postgres/collector/pg_activity.go

```diff
@@ -0,0 +1,373 @@
+		return err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		activity := ActivityInfo{}
+		err := rows.Scan(
+			&activity.DatabaseName,
+			&activity.DatabaseID,
+			&activity.PID,
+			&activity.LeaderPID,
+			&activity.UserSysID,
+			&activity.Username,
+			&activity.ApplicationName,
+			&activity.ClientAddr,
+			&activity.ClientPort,
+			&activity.StateChange,
+			&activity.Now,
+			&activity.BackendStart,
+			&activity.XactStart,
+			&activity.QueryStart,
+			&activity.WaitEventType,
+			&activity.WaitEvent,
+			&activity.State,
+			&activity.BackendType,
+			&activity.BackendXID,
+			&activity.BackendXmin,
+			&activity.QueryID,
+			&activity.Query,
+			&activity.BlockedByPids,
+		)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "failed to scan pg_stat_activity set", "err", err)
+			continue
+		}
+
+		_, err = c.validateActivity(activity)
+		if err != nil {
+			level.Debug(c.logger).Log("msg", "invalid pg_stat_activity set", "err", err)
+			continue
+		}
+
+		query := ""
+		if activity.Query.Valid {
+			query = activity.Query.String
+		}
+
+		leaderPID := ""
+		if activity.LeaderPID.Valid {
+			leaderPID = fmt.Sprintf(`%d`, activity.LeaderPID.Int64)
+		}
+
+		stateDuration := calculateDuration(activity.StateChange, activity.Now)
+		queryDuration := calculateDuration(activity.QueryStart, activity.Now)
+		xactDuration := calculateDuration(activity.XactStart, activity.Now)
+		backendDuration := calculateDuration(activity.BackendStart, activity.Now)
+
+		databaseName := ""
+		if activity.DatabaseName.Valid {
+			databaseName = activity.DatabaseName.String
+		}
+
+		userName := ""
+		if activity.Username.Valid {
+			userName = activity.Username.String
+		}
+
+		applicationName := ""
+		if activity.ApplicationName.Valid {
+			applicationName = activity.ApplicationName.String
+		}
+
+		state := ""
+		if activity.State.Valid {
+			state = activity.State.String
+		}
+
+		clientAddr := ""
+		if activity.ClientAddr.Valid {
+			clientAddr = activity.ClientAddr.String
+			if activity.ClientPort.Valid {
+				clientAddr = fmt.Sprintf("%s:%d", clientAddr, activity.ClientPort.Int32)
+			}
+		}
+
+		backendType := ""
+		if activity.BackendType.Valid {
+			backendType = activity.BackendType.String
+		}
+
+		waitEventFullName := ""
+		waitEvent := ""
+		waitEventType := ""
+		if activity.WaitEventType.Valid && activity.WaitEvent.Valid {
+			waitEventFullName = fmt.Sprintf("%s:%s", activity.WaitEventType.String, activity.WaitEvent.String)
+			waitEvent = activity.WaitEvent.String
+			waitEventType = activity.WaitEventType.String
+		}
+
+		// Build query sample entry
+		sampleLabels := fmt.Sprintf(
+			`clock_timestamp="%s" instance="%s" app="%s" client="%s" backend_type="%s" backend_time="%s" state="%s" pid="%d" leader_pid="%s" user="%s" userid="%d" datname="%s" datid="%d" xact_time="%s" xid="%d" xmin="%d" query_time="%s" queryid="%d" query="%s" engine="postgres"`,
+			activity.Now.Format(time.RFC3339Nano),
```

Comment:
What is `clock_timestamp` for? We can use `BuildLokiEntryWithTimestamp` to set the timestamp of the log to match when the event took place

Reply:
Huum ok, will look into that.